### PR TITLE
Refactor MessageActionBar w/ARIA Toolbar pattern

### PR DESCRIPTION
**Pull request by jugglinmike**
This is intended to resolve vector-im/riot-web#11316

Signed-off-by: Mike Pennisi mike@mikepennisi.com

### CODE REVIEW

File: src/components/views/messages/MessageActionBar.js [hunk 1]
```diff
@@ -105,14 +112,24 @@ export default class MessageActionBar extends React.PureComponent {
...
+        super(props);
+        this.elRef = React.createRef();
+        this.state = { focused: "options", buttonNames: [] };
+    }
+
     componentDidMount() {
         this.props.mxEvent.on("Event.decrypted", this.onDecrypted);
         this.props.mxEvent.on("Event.beforeRedaction", this.onBeforeRedaction);
+
+        this.onKeyDown = this.onKeyDown.bind(this);
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:125]: you don't need this as you used an arrow func which will have the right `this` context

File: src/components/views/messages/MessageActionBar.js [hunk 2]
```diff
@@ -149,45 +197,61 @@ export default class MessageActionBar extends React.PureComponent {
...
-        let replyButton;
-        let editButton;
+        const buttons = [];
 
         if (isContentActionable(this.props.mxEvent)) {
             if (this.context.canReact) {
-                reactButton = (
-                    <ReactButton mxEvent={this.props.mxEvent} reactions={this.props.reactions} onFocusChange={this.onFocusChange} />
+                buttons.push(
+                    <ReactButton mxEvent={this.props.mxEvent} reactions={this.props.reactions} onFocusChange={this.onFocusChange}
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:205]: can you style these props like the reply button below (one per row starting on row after opening

File: src/components/views/messages/MessageActionBar.js [hunk 3]
```diff
@@ -149,45 +197,61 @@ export default class MessageActionBar extends React.PureComponent {
...
+        );
+
+        const newButtonNames = buttons.map((button) => button.key);
+        if (this.state.buttonNames.join() !== newButtonNames.join()) {
+            this.setState({buttonNames: newButtonNames});
+        }
+
+        // aria-live=off to not have this read out automatically as navigating around timeline, gets repetitive.
+        return <div className="mx_MessageActionBar" role="toolbar" aria-label={_t("Message Actions")} aria-live="off"
+            ref={this.elRef}>
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:253]: same as above re props 1-per-line style please

File: src/components/views/messages/MessageActionBar.js (hunk 1 above)
Comment (jugglinmike) [reply src/components/views/messages/MessageActionBar.js:125]: Ah, got it. Still haven't really learned the semantics of class fields

File: src/components/views/messages/MessageActionBar.js (hunk 2 above)
Comment (jugglinmike) [reply src/components/views/messages/MessageActionBar.js:205]: Sure

**Comment by t3chguy** [comment]
Can you do a DCO sign-off in the PR description as one of your commits lacks one, thanks :)

File: src/components/views/messages/MessageActionBar.js [hunk 4]
```diff
@@ -133,6 +148,37 @@ export default class MessageActionBar extends React.PureComponent {
         this.props.onFocusChange(focused);
     };
 
+    onKeyDown = (ev) => {
+        const {key} = ev;
+        const buttonNames = this.state.buttonNames;
+        const current = buttonNames.indexOf(this.state.focused);
+        let newIndex = null;
+
+        if (key === Key.ARROW_UP || key === Key.ARROW_DOWN) {
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:157]: Hmm, this seems to disagree with https://www.w3.org/TR/wai-aria-practices-1.1/examples/toolbar/toolbar.html

**Comment by jugglinmike** [comment]
> Hmm, this seems to disagree with https://www.w3.org/TR/wai-aria-practices-1.1/examples/toolbar/toolbar.html

Looks like the "up" and "down" arrows should only perform a click for targets with `aria-haspopup`. In other words: those keys should have no effect for "Reply" and "Edit." Is that what you had in mind?

**Comment by t3chguy** [comment]
So the codebase already has some classes to help out with keeping track of RovingTabIndex. It'd be nice to use that and make a generic Toolbar component which would handle the keypresses between all of its direct children so that we can have one backing solution for both this and https://github.com/matrix-org/matrix-react-sdk/pull/4799

> In other words: those keys should have no effect for "Reply" and "Edit." Is that what you had in mind?

Yes, that's what the aria authoring practices say

**Comment by jugglinmike** [comment]
Are you saying we should do that work before implementing this feature?

**Comment by t3chguy** [comment]
I am saying it'd be nice to write a component which does this for all Toolbars rather than manually having to add these keyboard handlers to every toolbar component manually, means a lot more repeated work

**Comment by jugglinmike** [comment]
Totally! I just can't tell if you consider that improvement to be blocking this one.

**Comment by t3chguy** [comment]
It'd be a replacement which would replace this one, so whilst this could land, it'd be replaced by this generic toolbar thing

File: src/components/views/messages/MessageActionBar.js [hunk 5]
```diff
@@ -105,14 +112,22 @@ export default class MessageActionBar extends React.PureComponent {
...
+    constructor(props) {
+        super(props);
+        this.elRef = React.createRef();
+        this.state = { focused: "options", buttonNames: [] };
+    }
+
     componentDidMount() {
         this.props.mxEvent.on("Event.decrypted", this.onDecrypted);
         this.props.mxEvent.on("Event.beforeRedaction", this.onBeforeRedaction);
+        this.elRef.current.addEventListener("keydown", this.onKeyDown, true);
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:124]: we should really be using react keydown handlers for this

(note to self mostly, as I am trying to make a generic Toolbar component now)

**Pull request closed without merging**